### ISSUE TITLE
Can `collapse::collap()` skip groups with size 1?

### ISSUE THREAD

**Issue by emmansh** [SebKrantz/collapse#193]
<sup>Created on 2021-10-31 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>
When dealing with grouped data, can we apply `collapse::collap()` only on groups in which `fndistinct > 1`?
I have grouped data with > 3M rows, and I need to collapse it by `mean()`. Since there are many groups, the procedure is taking quite some time. When investigating the data, I realized that most of the groups are in fact of size 1; meaning, they don't require any collapsing. So I need to iterate only over a small portion of groups when collapsing. I wonder whether being able to apply `collap()` only over groups with `fndistinct > 1` would cut down execution time substantially. 

Reproducible example
---
In the following data we have 3 columns:
- `ids`
- `years`
- `value`
I want to calculate the mean of `value` column when the data is grouped by `ids` & `years`. That is, for every distinct combination of  `ids` & `years`, I want one `value`.

``` r
library(ids)
library(magrittr)
library(tibble)
library(tidyr, warn.conflicts = FALSE)
library(collapse, warn.conflicts = FALSE)
#> collapse 1.6.5, see ?`collapse-package` or ?`collapse-documentation`
#> Note: stats::D  ->  D.expression, D.call, D.name
library(tictoc)

generate_random_unique_ids <- function(n) {
  vec_ids <- ids::random_id(n = n, bytes = 4, use_openssl = FALSE)
  repeat {
    duplicates <- duplicated(vec_ids)
    if (!any(duplicates)) {
      break
    }
    vec_ids[duplicates] <- ids::random_id(n = sum(duplicates), bytes = 4, use_openssl = FALSE)
  }
  vec_ids
}

set.seed(2021)

ids              <- generate_random_unique_ids(4e6)
years            <- sample(1930:2022, size = 4e6, replace = TRUE)
id_yr_group_size <- rep(c(1, 4), times = c(4e6-5000, 5000)) 

my_df <-
  tibble::tibble(ids, years, id_yr_group_size) %>%
  tidyr::uncount(id_yr_group_size) %>%
  collapse::ftransform(value = rnorm(n = nrow(.)))
```
The special thing about `my_df` is that its grouping almost means nothing, as only 20,000 rows (out of 4M) belong to more than 1 combination of `ids` & `years` (due to expanding rows according to weights from `id_yr_group_size `).
```r
GRP(my_df, ~ ids + years)
#> collapse grouping object of length 4015000 with 4000000 ordered groups
#> 
#> Call: GRP.default(X = my_df, by = ~ids + years), X is unordered
#> 
#> Distribution of group sizes: 
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>   1.000   1.000   1.000   1.004   1.000   4.000 
#> 
#> Groups with sizes: 
#> 000001cd.1987 00000235.1972 0000029d.1975 00000348.1981 00000362.1957 
#>             1             1             1             1             1 
#> 0000046d.1960 
#>             1 
#>   ---
#> fffff405.1953 fffff541.1936 fffff80b.2014 fffff896.1990 fffffe88.1950 
#>             1             1             1             1             1 
#> fffffe96.1952 
#>             1
```
But when we run `collap()` it seems to iterate over **all** the groups, even though there's no per-group aggregation needed in most of them.
```r
tic()
my_df_collapsed_everything <- collap(my_df, ~ ids + years, FUN = mean)
toc()
#> 38.04 sec elapsed
```
Had `collap()` skipped the groups that are of size 1, and running only over those that actually need aggregation, the entire procedure would have been many times speedier.
```r
tic()
my_df_collapsed_bottom_20k <- 
  my_df %>%
  tail(20000) %>%
  collap(., ~ ids + years, FUN = mean)
toc()
#> 0.02 sec elapsed
```
Obviously, there are workarounds on the user's end, such as splitting the data to two objects, according to group sizes, applying and finally binding back. But wouldn't it be nice if `collap()` had silently skipped rows where there's nothing to collapse?

**Comment by SebKrantz** [comment]
Thanks, well first of all I would like to encourage you to really read the documentation of `collap` and also the vignettes. Then you would notice that this package provides fast statistical functions that are optimized for big data computations `?.FAST_STAT_FUN`. All examples I give everywhere, including on `?collap` use these fast functions. The virtue of these fast functions is that they compute statistics in parallel over groups. Read [this](https://sebkrantz.github.io/collapse/articles/collapse_and_dplyr.html#excursus-what-is-happening-behind-the-scenes) small section to get the idea. So all you need to do here is put `FUN = fmean`. Notice that most of the time in your example data is spent on grouping not on calculating the mean:
```r
system.time(GRP(my_df, ~ ids + years)) # 1 sec
system.time(collap(my_df, ~ ids + years, FUN = fmean)) # 1.2 sec
```
Of course applying base R's mean over 4 million groups is extremely slow, and that is because this pacakge was not built to apply base R functions efficiently to anything. `dplyr`is much more efficient that this package in doing that. 
Identifying singleton groups is also possible using the functions in this package, albeit a bit more advanced: 
```r
o <- with(my_df, radixorder(ids, years, starts = TRUE, group.sizes = TRUE))
single_group_idx <- o[attr(o, "starts")[attr(o, "group.sizes") == 1L]]
my_df_ag <- rbind(ss(my_df, single_group_idx), collap(ss(my_df, -single_group_idx), ~ ids + years))
```
In this case I don't think it gives a performance increase, as you are effictively grouping the data twice, and also subsetting 2 times and rbinding, which are also quite expensive operations. 

So bottom line: Use the fast statistical functions where you can, and examine the documentaion before creating an issue.

**Comment by emmansh** [comment]
@SebKrantz I see. Thank you for the response.
